ksql-engine/src/test/java/io/confluent/ksql/planner/plan/KsqlStructuredDataOutputNodeTest.java

```diff
@@ -270,86 +272,132 @@ public void shouldComputeQueryIdCorrectlyForInsertInto() {
         queryContextCaptor.capture()
     );
 
-    assertThat(QueryLoggerUtil.queryLoggerName(queryContextCaptor.getValue()), is("output-test.0"));
+    assertThat(QueryLoggerUtil.queryLoggerName(queryContextCaptor.getValue()),
+        is("output-test.0"));
   }
 
-  private <K> KsqlStructuredDataOutputNode getKsqlStructuredDataOutputNodeForTable(
-      final SerdeFactory<K> keySerdeFatory
+  @Test
+  public void shouldCallInto() {
+    // When:
+    outputNode.buildStream(ksqlStreamBuilder);
+
+    // Then:
+    verify(resultStream).into(
+        SINK_KAFKA_TOPIC_NAME,
+        rowSerde,
+        ImmutableSet.of()
+    );
+  }
+
+  @Test
+  public void shouldCallIntoWithIndexesToRemoveImplicitsAndRowKey() {
+    // Given:
+    final LogicalSchema schema = SCHEMA.withImplicitFields();
+    createOutputNode(schema);
+
+    // When:
+    outputNode.buildStream(ksqlStreamBuilder);
+
+    // Then:
+    verify(resultStream).into(
+        SINK_KAFKA_TOPIC_NAME,
+        rowSerde,
+        ImmutableSet.of(0, 1)
+    );
+  }
+
+  @Test
+  public void shouldCallIntoWithIndexesToRemoveImplicitsAndRowKeyRegardlessOfLocation() {
+    // Given:
+    final LogicalSchema schema = LogicalSchema.of(SchemaBuilder.struct()
+        .field("field1", Schema.OPTIONAL_STRING_SCHEMA)
+        .field("field2", Schema.OPTIONAL_STRING_SCHEMA)
+        .field("ROWKEY", Schema.OPTIONAL_STRING_SCHEMA)
+        .field("field3", Schema.OPTIONAL_STRING_SCHEMA)
+        .field("timestamp", Schema.OPTIONAL_INT64_SCHEMA)
+        .field("ROWTIME", Schema.OPTIONAL_INT64_SCHEMA)
+        .field("key", Schema.OPTIONAL_STRING_SCHEMA)
+        .build());
+
+    createOutputNode(schema);
+
+    // When:
+    outputNode.buildStream(ksqlStreamBuilder);
+
+    // Then:
+    verify(resultStream).into(
+        SINK_KAFKA_TOPIC_NAME,
+        rowSerde,
+        ImmutableSet.of(2, 5)
+    );
+  }
+
+  private void createOutputNode(
+      final boolean partitionBy,
+      final boolean createInto
   ) {
-    final KsqlTable<K> dataSource = new KsqlTable<>(
-        "sqlExpression", "datasource",
-        schema,
+    outputNode = new KsqlStructuredDataOutputNode(
+        PLAN_NODE_ID,
+        sourceNode,
+        SCHEMA,
+        new LongColumnTimestampExtractionPolicy("timestamp"),
+        KEY_FIELD,
+        ksqlTopic,
+        partitionBy,
+        OptionalInt.empty(),
+        createInto,
         SerdeOption.none(),
-        KeyField.of("key", schema.getSchema().field("key")),
-        new MetadataTimestampExtractionPolicy(),
-        new KsqlTopic(SOURCE_TOPIC_NAME, SOURCE_KAFKA_TOPIC_NAME, new KsqlJsonSerdeFactory(),
-            false),
-        keySerdeFatory
+        sinkFactory
     );
+  }
 
-    final DataSourceNode tableSourceNode = new DataSourceNode(
-        new PlanNodeId("0"),
-        dataSource,
-        dataSource.getName());
-
-    return new KsqlStructuredDataOutputNode(
-        new PlanNodeId("0"),
-        tableSourceNode,
+  private void createOutputNode(
```

Comment:
nit: can combine this and the method above (add the parameters from one to the other)

Reply:
This was deliberate. I prefer the shorter, more targeted versions.  Makes the tests read better.